protocol/contracts/TrucoMatch.sol

```diff
@@ -10,42 +10,132 @@ import "./crypt/DeckCrypt.sol";
+    }
+
+    // Method for staking & starting the match
+    function stake(uint8 _player) public {
+        require(
+            currentMatch.players[_player].playerAddress == msg.sender, 
+            "Player is not joined");
+        require(
+            truCoin.balanceOf(msg.sender) >= currentMatch.tokensAtStake, 
+            "You don't have enough Trucoins to stake this match");
+        require(
+            truCoin.allowance(msg.sender, address(this)) >= currentMatch.tokensAtStake, 
+            "Not enough trucoins transfer approved");
+        require(
+            currentMatch.players[_player].tokensAtStake == 0, 
+            "You already staked this match");
+
+        // Transfer Trucoins from player to contract
+        require(truCoin.transferFrom(
+            msg.sender, address(this), currentMatch.tokensAtStake), 
+            "Trucoin transfer failed");
+        
+        currentMatch.players[_player].tokensAtStake = currentMatch.tokensAtStake;
+
+        // emit event
+        emit PlayerStaked(msg.sender, currentMatch.tokensAtStake);
+
+        // Start match
+        if (currentMatch.players[0].tokensAtStake == currentMatch.tokensAtStake && 
+            currentMatch.players[1].tokensAtStake == currentMatch.tokensAtStake) {
+            currentMatch.gameState = trucoEngine.startGame();
+            emit MatchStarted(
+                currentMatch.players[0].playerAddress, 
+                currentMatch.players[1].playerAddress,
+                currentMatch.tokensAtStake);
+        }
     }
 
     function newDeal(CardsStructs.Deck memory _deck) public {
         // Check if current game state enables new card shufflings
-        // Check for player authorization to shuffle
+        require(!currentMatch.gameState.isDealOpen, "Deal is already open");
+
+        // Determine the new shuffler and check that corresponds to the current player
+        uint8 new_shuffler = currentMatch.gameState.playerWhoShuffled ^ 1;
+        require(msg.sender == currentMatch.players[new_shuffler].playerAddress, "You are not the shuffler");
+
         currentMatch.deck = _deck; // encrypted deck using 2 secrets
-        // Assign cards for each player interleaving one for each other, i.e player1: 0,2,4 player2: 1,3,5
+        
+        // Assign new shuffler
+        currentMatch.gameState.playerWhoShuffled = new_shuffler;
+
+        // Assign turn to player not shuffling
+        currentMatch.gameState.playerTurn = currentMatch.gameState.playerWhoShuffled ^ 1;
+
+        // Unset current challenge from previous deal
+        currentMatch.gameState.currentChallenge.challenge = CardsStructs.Challenge.None;
+
+        // Clean revealed cards
+        currentMatch.gameState.revealedCardsByPlayer[0] = [0,0,0];
+        currentMatch.gameState.revealedCardsByPlayer[1] = [0,0,0];
+
+        // Clean envido count
+        currentMatch.gameState.envidoCountPerPlayer[0] = 0;
+        currentMatch.gameState.envidoCountPerPlayer[1] = 0;
```

Comment:
This structure will change, but it's ok since we should aim at mergin this PR as soon as possible and i will fix it later on